File: Python_Code/11.1.1_Thermometer/Thermometer3.py
import time
import curses
import pigpio
import math

YL_40=0x48
pi = pigpio.pi() # Connect to local Pi.
handle = pi.i2c_open(1, YL_40, 0)
 
stdscr = curses.initscr()
curses.noecho()
curses.cbreak()
aout = 0

stdscr.addstr(10, 0, "Temperature")
# stdscr.addstr(12, 0, "Temperature")
# stdscr.addstr(14, 0, "AOUT->AIN2")
# stdscr.addstr(16, 0, "Resistor")

stdscr.nodelay(1)

while True:
    for a in range(0,4):
        aout = aout + 1
        #pi.i2c_write_byte_data(handle, 0x40 | ((a+1) & 0x03), aout&0xFF)
        pi.i2c_write_byte_data(handle, 0x40 | (0x07), aout&0xFF) #connected to AIN3 at addsress 0x07
        v = pi.i2c_read_byte(handle)
        
        voltage = v / 255.0 * 3.3        # calculate voltage
        Rt = 10 * voltage / (3.3 - voltage)    # calculate resistance value of thermistor
        tempK = 1/(1/(273.15 + 25) + math.log(Rt/10)/3950.0) # calculate temperature (Kelvin)
        tempC = tempK -273.15        # calculate temperature (Celsius)
        string = ('ADC Value : %d, Voltage : %.2f, Temperature : %.2f'%(v,voltage,tempC))

        stdscr.addstr(10, 12, str(string) + ' ')

    stdscr.refresh()
    time.sleep(0.04)
    pi.set_PWM_dutycycle(17, v) 
    c = stdscr.getch()

    if c != curses.ERR:
        break

curses.nocbreak()
curses.echo()
curses.endwin()

pi.i2c_close(handle)
pi.write(17,0)
pi.stop()
print("bye")

File: C_Code/11.1.1_Thermometer/Thermometer3.c
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <pigpio.h>
#include <ncurses.h> 
#include <math.h>

#define PCF8591_I2C_ADDR 0x48

int main(int argc, char *argv[])
{
    int i;
    int r;
   int handle;
   char aout;
   char command[2];
   // unsigned char value[4];
   unsigned char str[8];
   int value;

   int j;
   int key;

   if (gpioInitialise() < 0) return 1;

   initscr();
   noecho();
   cbreak();
   nodelay(stdscr, true);
   curs_set(0);

   printw("PCF8591  - press any key to quit.");

   mvaddstr(10, 0, "Thermometer");
   refresh();

   handle = i2cOpen(1, PCF8591_I2C_ADDR, 0);

   command[1] = 0;
   aout = 128;
   float voltage;

   while (1)
   {
        command[1] = aout;
        command[0] = 0x40 | (0x07); // output enable | read input i

        i2cWriteDevice(handle, &command, 2);
        value = i2cReadByte(handle);
        voltage = value / 255.0 * 3.3;
        float Rt = 10 * voltage / (3.3 - voltage); //calculate resistance value of thermistor
        float tempK = 1/(1/(273.15 + 25) + log(Rt/10)/3950.0); //calculate temperature (Kelvin)
        float tempC = tempK - 273.15; //calculate temperature (Celsius)

        sprintf(str, "%3d, %f, temp %.1fC", value, voltage, tempC);
    
        mvaddstr(10 , 18, str);
        refresh();
        gpioPWM(17, value); //  just one line to get the LED on
        key = getch();
        if (key != -1)
        break;
   }
   endwin();
   i2cClose(handle);
   gpioWrite(17,1);
   gpioTerminate();
   return (0);
}
